### DISCUSSION TITLE
Reverts not reducung

### DISCUSSION THREAD

**Discussion by 0x-codeX** [Cyfrin/foundry-full-course-cu#1617] [Q&A] [answered]
Each time I run my Invariant forge test with fail_on_revert set to false, the revert number remains the same with calls, I noticed it started after I added these two lines in my handler contract `handler = new Handler(dsce, dsc);
        targetContract(address(handler));`

and deleted this 
`targetContract(address(dsce));`

Invariants.t.sol

`
// SPDX-License-Identifier: MIT

// This Invariant file will have our invariants aka properties of the system that should always hold. 

//What are our invariants?

//1. The total supply of DSC should be less than the total value of the collateral. 
//2. Getter view functions should never revert. 

pragma solidity 0.8.20;

import { Test } from "forge-std/Test.sol";
import { StdInvariant } from "forge-std/StdInvariant.sol";
import { DeployDSC } from "../../../script/DeployDSC.s.sol";
import { DSCEngine } from "../../../src/DSCEngine.sol";
import { DecentralizedStableCoin } from "../../../src/DecentralizedStableCoin.sol";
import { HelperConfig } from "../../../script/HelperConfig.s.sol";
import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
import { console } from "forge-std/Test.sol";
import { Handler } from "./Handler.t.sol";

contract InvariantsTest is StdInvariant, Test {
    DeployDSC deployer;
    DSCEngine dsce;
    DecentralizedStableCoin dsc;
    HelperConfig config;
    address weth;
    address wbtc;
    Handler handler;

    function setUp() external {
        deployer = new DeployDSC();
        (dsc, dsce, config) = deployer.run();
        (,, weth, wbtc,) = config.activeNetworkConfig();
        //targetContract(address(dsce));
        handler = new Handler(dsce, dsc);
        targetContract(address(handler));
    }

    function invariant_protocolMustHaveMoreValueThanTotalSupply() public view {
      //get the value of all the collateral in the protocol then,
      //compare it to all the debt(dsc)
      uint256 totalSupply = dsc.totalSupply(); // total supply of all dsc in the world.
      uint256 totalWethDeposited = IERC20(weth).balanceOf(address(dsce)); //total amount of weth deposited into the SC.
      uint256 totalWbtcDeposited = IERC20(wbtc).balanceOf(address(dsce)); //total amount of wbtc Deposit into the SC.
 
      uint256 wethValue = dsce.getUsdValue(weth, totalWethDeposited);
      uint256 wbtcValue = dsce.getUsdValue(wbtc, totalWbtcDeposited);

      console.log("weth value: ", wethValue);
      console.log("wbtc value: ", wbtcValue);
      console.log("total supply: ", totalSupply);

      assert(wethValue + wbtcValue >= totalSupply);
    }
}

`

Error message

`
Ran 1 test for test/unit/fuzz/Invariants.t.sol:OpenInvariantsTest
[PASS] invariant_protocolMustHaveMoreValueThanTotalSupply() (runs: 128, calls: 16384, reverts: 16384)
Test result: ok. 1 passed; 0 failed; 0 skipped; finished in 1.80s

`

I cannot see what I missed, thank you.

**Comment by EngrPips** [comment, marked as answer]
Hello @codexNature, whenever your total number of reverted calls equals the total number of calls, all calls made to the target contract reverted, and none of the calls were reasonable. That would always happen when you are testing a contract with many states and criteria that need to be met to make a successful call to some function of the contract.

That is the work of a `handler` contract in a `controlled stateful fuzzing` test suite. So, please ensure your handler properly ensures each call to the target contract is reasonable. I would suggest you see Patrick's test suite `https://github.com/Cyfrin/foundry-defi-stablecoin-f23/tree/main/test/fuzz` to see what you are doing wrongly.

**Reply by 0x-codeX** [reply]
Thank you once again @EngrPips I did as you suggested and I was even able to get my reverts down to 0 just like Patrick did.

**Reply by EngrPips** [reply]
Awesome, I am glad you figured it out. keep crushing it, friend.

**Reply by FrankezeCode** [reply]
@codexNature bro , how did you solve the problem, ? What exactly is causing the Reverts Not reducing.  I'm having the same issue. 
I understand yours went to Zero finally, but was it actually reducing, did you notice any reduction before the final Zero?.

**Reply by EngrPips** [reply]
Maybe visit https://github.com/Cyfrin/foundry-defi-stablecoin-f23/tree/main/test/fuzz to see how our Teacher Patrick did it.
